### PR TITLE
Code server redesign

### PR DESCRIPTION
**Pull request by ankitjavalkar**

### CODE REVIEW

File: testapp/exam/admin.py [hunk 1]
```diff
@@ -1,5 +1,6 @@
-from testapp.exam.models import Question, Quiz
+from exam.models import Question, Quiz, TestCase
```
Comment (prabhuramachandran) [inline testapp/exam/admin.py]: Why is this changed?

File: testapp/exam/code_server.py [hunk 2]
```diff
@@ -53,167 +54,311 @@ def timeout_handler(signum, frame):
     """A handler for the ALARM signal."""
     raise TimeoutException('Code took too long to run.')
 
+def create_delete_signal_handler(action=None, old_handler=None):
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: Split this into three different functions with clear descriptive names as discussed.

File: testapp/exam/code_server.py [hunk 3]
```diff
@@ -53,167 +54,311 @@ def timeout_handler(signum, frame):
...
-    """
-    def __init__(self, port, queue):
-        self.port = port
-        self.queue = queue
-        msg = 'Code took more than %s seconds to run. You probably '\
-              'have an infinite loop in your code.' % SERVER_TIMEOUT
-        self.timeout_msg = msg
+class TestCode(object):
+    """Evaluates and tests the code obtained from Code Server"""
+    def __init__(self, info_parameter, in_dir=None):
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: Fully document args, use parameter_info_json if you expect json.

File: testapp/exam/code_server.py [hunk 4]
```diff
@@ -53,167 +54,311 @@ def timeout_handler(signum, frame):
...
+        info_parameter = json.loads(info_parameter)
+        self.test_parameter = info_parameter.get("test_parameter")
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: It seems you might as well pass kwargs to the `__init__` above instead of json.

File: testapp/exam/code_server.py [hunk 5]
```diff
@@ -53,167 +54,311 @@ def timeout_handler(signum, frame):
...
+        self.language = info_parameter.get("language")
+        self.user_answer = info_parameter.get("user_answer")
+        self.in_dir = in_dir
+
+    def run_code(self):
+        """Tests given code  (`answer`) with the `test_code` supplied.
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: Fix docstring, there is no test code supplied here...

File: testapp/exam/code_server.py [hunk 6]
```diff
@@ -53,167 +54,311 @@ def timeout_handler(signum, frame):
...
+
+        if self.language == "python":
+            tb = None
+            test_code = self.create_test_case()
+            # Add a new signal handler for the execution of this code.
+            prev_handler = create_delete_signal_handler("new")
+
+            try:
+                submitted = compile(self.user_answer, '<string>', mode='exec')
+                g = {}
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: Refactor as discussed.

File: testapp/exam/code_server.py [hunk 7]
```diff
@@ -53,167 +54,311 @@ def timeout_handler(signum, frame):
...
-            type, value = sys.exc_info()[:2]
-            err = "Error: {0}".format(repr(value))
         else:
-            success = True
-            err = 'Correct answer'
-        finally:
-            del tb
-            # Set back any original signal handler.
-            signal.signal(signal.SIGALRM, old_handler)
+            user_answer_file = {'C': 'submit.c', 'java': 'Test.java', 'scilab': 'function.sci',
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: Use subclasses to segregate the different languages.  A method more than a screenful should be ruthlessly refactored.

File: testapp/exam/code_server.py [hunk 8]
```diff
@@ -53,399 +54,46 @@ def timeout_handler(signum, frame):
...
+    return
+
+
+
+###############################################################################
+# `TestCode` class.
+###############################################################################
+class TestCode(object):
+    """Evaluates and tests the code obtained from Code Server"""
+    def __init__(self, test_parameter, language, user_answer, ref_code_path=None, in_dir=None):
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: Wrap lines at 80 cols, follow pep8.

File: testapp/exam/code_server.py [hunk 9]
```diff
@@ -461,29 +109,29 @@ def run_cplus_code(self, answer, test_code, in_dir=None):
...
-        submit_f.close()
-        submit_path = abspath(submit_f.name)
-
-        ref_path = test_code.strip()
-        if not ref_path.startswith('/'):
-            ref_path = join(MY_DIR, ref_path)
+        success = False
+        prev_handler = None
+
+        methods = {"python": 'evaluate_python_code',
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: This is not good.  It means we'll have to change this method each time a bloke wants to support a new language.  Just use a subclass where the subclass implements the appropriate `evaluate` method.  You can have perhaps a classmethod called "register" which can be called by subclasses to register their new language implementations.

File: testapp/exam/code_server.py [hunk 10]
```diff
@@ -461,29 +109,29 @@ def run_cplus_code(self, answer, test_code, in_dir=None):
...
-        old_handler = signal.signal(signal.SIGALRM, timeout_handler)
-        signal.alarm(SERVER_TIMEOUT)
+        prev_handler = create_signal_handler()
+        success = False
 
         # Do whatever testing needed.
-        success = False
         try:
-            success, err = self._check_c_cpp_code(ref_path, submit_path)
+            evaluate_code = getattr(self, get_method_based_on_lang)
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: Much nicer than before.

File: testapp/exam/code_server.py [hunk 11]
```diff
@@ -649,106 +363,131 @@ def _check_java_code(self, ref_code_path, submit_code_path):
...
-
-        The testcode is a path to the reference code.
-        The reference code will call the function submitted by the student.
-        The reference code will check for the expected output.
+        return success, err
 
-        If the path's start with a "/" then we assume they are absolute paths.
-        If not, we assume they are relative paths w.r.t. the location of this
-        code_server script.
+    def check_bash_script(self, ref_path, submit_path,
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: See this should be in a separate class and not muddle the basic `TestCode` class.

File: testapp/exam/code_server.py [hunk 12]
```diff
@@ -587,36 +272,66 @@ def _check_java_code(self, ref_code_path, submit_code_path):
...
+
+        language_dependent_path = {
+            'c_user_output_path': os.getcwd() +  '/output',
+            'c_ref_output_path': os.getcwd() + '/executable',
+            'java_student_directory': os.getcwd() + '/',
+            # 'java_student_file_name': 'Test',
+            'java_ref_file_name': (ref_code_path.split('/')[-1]).split('.')[0],
+        }
+
+        language_dependent_var = {
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: YIKES!

**Comment by prabhuramachandran** [comment]
Still needs a lot of work but is better than before.  Where are the tests?

File: testapp/exam/evaluate_bash_code.py [hunk 13]
```diff
@@ -0,0 +1,119 @@
...
+from os.path import join, isfile
+import subprocess
+import importlib
+
+# local imports
+from evaluate_code import EvaluateCode
+from language_registry import registry
+
+
+class EvaluateBashCode(EvaluateCode):
```
Comment (prabhuramachandran) [inline testapp/exam/evaluate_bash_code.py]: Make this name a noun and not a verb.

File: testapp/exam/evaluate_c_code.py [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+import importlib
+
+# local imports
+from evaluate_code import EvaluateCode
+from language_registry import registry
+
+
+class EvaluateCCode(EvaluateCode):
+    """Tests the C code obtained from Code Server"""
+    # Public Protocol ##########
```
Comment (prabhuramachandran) [inline testapp/exam/evaluate_c_code.py]: Keep one for the entire public API, not one for each function!

File: testapp/docs/sample_questions.py [hunk 15]
```diff
@@ -1,84 +0,0 @@
-from datetime import date
```
Comment (prabhuramachandran) [inline testapp/docs/sample_questions.py]: Why is this removed?

File: testapp/exam/evaluate_code.py [hunk 16]
```diff
@@ -0,0 +1,191 @@
...
+        # Cancel the signal
+        delete_signal_handler()
+
+        result = {'success': success, 'error': err}
+        return result
+
+    def evaluate_code(self):
+        raise NotImplementedError("evaluate_code method not implemented")
+
+    def create_submit_code_file(self, file_name):
```
Comment (prabhuramachandran) [inline testapp/exam/evaluate_code.py]: These are not public methods.

File: testapp/exam/evaluate_code.py [hunk 17]
```diff
@@ -0,0 +1,191 @@
...
+        """ Write the code (`answer`) to a file and set the file path"""
+        submit_f = open(file_name, 'w')
+        submit_f.write(self.user_answer.lstrip())
+        submit_f.close()
+        submit_path = abspath(submit_f.name)
+
+        return submit_path
+
+    def set_file_as_executable(self, fname):
```
Comment (prabhuramachandran) [inline testapp/exam/evaluate_code.py]: These are all internal details so should be private.

File: testapp/exam/language_registry.py [hunk 18]
```diff
@@ -0,0 +1,17 @@
...
+    # Public Protocol ##########
+    def get_class(self, language):
+        return self._registry[language]
+
+    # Public Protocol ##########
+    def register(self, language, cls):
+        self._registry[language] = cls
+
+
+registry = LanguageRegistry()
```
Comment (prabhuramachandran) [inline testapp/exam/language_registry.py]: Add set_registry and get_registry functions here to make it more explicit.

File: testapp/exam/code_evaluator.py [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+        raise Exception("Signal Handler: object cannot be NoneType")
+
+
+def delete_signal_handler():
+    signal.alarm(0)
+    return
+
+
+###############################################################################
+# `TestCode` class.
```
Comment (prabhuramachandran) [inline testapp/exam/code_evaluator.py]: Remove this comment.

File: testapp/exam/code_evaluator.py [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+
+
+# Raised when the code times-out.
+# c.f. http://pguides.net/python/timeout-a-function
+class TimeoutException(Exception):
+    pass
+
+
+# Private Protocol ##########
+def timeout_handler(signum, frame):
```
Comment (prabhuramachandran) [inline testapp/exam/code_evaluator.py]: `_timeout_handler`?  Remove private protocol above.

File: testapp/exam/code_evaluator.py [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+        return success, err
+
+    def _teardown(self):
+        # Cancel the signal
+        delete_signal_handler()
+
+    def _check_code(self):
+        raise NotImplementedError("check_code method not implemented")
+
+    def create_submit_code_file(self, file_name):
```
Comment (prabhuramachandran) [inline testapp/exam/code_evaluator.py]: Why are these not prefixed with an underscore if they are private?

File: testapp/exam/code_server.py [hunk 22]
```diff
@@ -29,12 +28,22 @@
 from multiprocessing import Process, Queue
 import subprocess
 import re
+import json
 # Local imports.
 from settings import SERVER_PORTS, SERVER_TIMEOUT, SERVER_POOL_PORT
+from language_registry import set_registry
+# from evaluate_python_code import EvaluatePythonCode
```
Comment (prabhuramachandran) [inline testapp/exam/code_server.py]: Delete all commented code.

File: testapp/exam/cpp_code_evaluator.py [hunk 23]
```diff
@@ -0,0 +1,126 @@
...
+        compile_command = 'g++  {0} -c -o {1}'.format(self.submit_path,
+                                                 c_user_output_path)
+        compile_main = 'g++ {0} {1} -o {2}'.format(ref_path,
+                                                c_user_output_path,
+                                                c_ref_output_path)
+        run_command_args = [c_ref_output_path]
+        remove_user_output = c_user_output_path
+        remove_ref_output = c_ref_output_path
+
+        return ref_path, self.submit_path, compile_command, compile_main, run_command_args, remove_user_output, remove_ref_output
```
Comment (prabhuramachandran) [inline testapp/exam/cpp_code_evaluator.py]: 80 col limit?

File: testapp/exam/cpp_code_evaluator.py [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+        proc, stdnt_stderr = ret
+        # if self.language == "java":
+        stdnt_stderr = self._remove_null_substitute_char(stdnt_stderr)
+
+        # Only if compilation is successful, the program is executed
+        # And tested with testcases
+        if stdnt_stderr == '':
+            ret = self._compile_command(compile_main)
+            proc, main_err = ret
+            # if self.language == "java":
```
Comment (prabhuramachandran) [inline testapp/exam/cpp_code_evaluator.py]: remove commented code.

File: testapp/exam/cpp_code_evaluator.py [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+        if not isfile(ref_code_path):
+            return False, "No file at %s or Incorrect path" % ref_code_path
+        if not isfile(submit_code_path):
+            return False, 'No file at %s or Incorrect path' % submit_code_path
+
+        success = False
+        # output_path = os.getcwd() + '/output'
+        ret = self._compile_command(compile_command)
+        proc, stdnt_stderr = ret
+        # if self.language == "java":
```
Comment (prabhuramachandran) [inline testapp/exam/cpp_code_evaluator.py]: remove

File: testapp/exam/cpp_code_evaluator.py [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+        if the required permissions are not given to the file(s).
+
+        """
+        if not isfile(ref_code_path):
+            return False, "No file at %s or Incorrect path" % ref_code_path
+        if not isfile(submit_code_path):
+            return False, 'No file at %s or Incorrect path' % submit_code_path
+
+        success = False
+        # output_path = os.getcwd() + '/output'
```
Comment (prabhuramachandran) [inline testapp/exam/cpp_code_evaluator.py]: remove

File: testapp/exam/java_code_evaluator.py [hunk 27]
```diff
@@ -0,0 +1,128 @@
...
+        if not isfile(ref_code_path):
+            return False, "No file at %s or Incorrect path" % ref_code_path
+        if not isfile(submit_code_path):
+            return False, 'No file at %s or Incorrect path' % submit_code_path
+
+        success = False
+        # output_path = os.getcwd() + '/output'
+        ret = self._compile_command(compile_command)
+        proc, stdnt_stderr = ret
+        # if self.language == "java":
```
Comment (prabhuramachandran) [inline testapp/exam/java_code_evaluator.py]: remove

File: testapp/exam/java_code_evaluator.py [hunk 28]
```diff
@@ -0,0 +1,128 @@
...
+        stdnt_stderr = self._remove_null_substitute_char(stdnt_stderr)
+
+        # Only if compilation is successful, the program is executed
+        # And tested with testcases
+        if stdnt_stderr == '':
+            ret = self._compile_command(compile_main)
+            proc, main_err = ret
+            # if self.language == "java":
```
Comment (prabhuramachandran) [inline testapp/exam/java_code_evaluator.py]: remove

File: testapp/exam/models.py [hunk 29]
```diff
@@ -20,7 +21,7 @@ class Profile(models.Model):
         ("python", "Python"),
         ("bash", "Bash"),
         ("C", "C Language"),
-        ("C++", "C++ Language"),
+        ("CPP", "C++ Language"),
```
Comment (prabhuramachandran) [inline testapp/exam/models.py]: Why are C and CPP alone upper case and the rest lower case?  Make them all lowercase.

File: testapp/exam/python_code_evaluator.py [hunk 30]
```diff
@@ -0,0 +1,58 @@
...
+        elif self.test_case_data:
+            for test_case in self.test_case_data:
+                pos_args = ", ".join(str(i) for i in test_case.get('pos_args')) \
+                                    if test_case.get('pos_args') else ""
+                kw_args = ", ".join(str(k+"="+a) for k, a
+                                 in test_case.get('kw_args').iteritems()) \
+                                if test_case.get('kw_args') else ""
+                args = pos_args + ", " + kw_args if pos_args and kw_args \
+                                                    else pos_args or kw_args
+                tcode = "assert {0}({1}) == {2}".format(test_case.get('func_name'),
```
Comment (prabhuramachandran) [inline testapp/exam/python_code_evaluator.py]: Formatting

File: testapp/exam/settings.py [hunk 31]
```diff
@@ -18,3 +18,11 @@
 # reason set this to the root you have to serve at.  In the above example
 # host.org/foo/exam set URL_ROOT='/foo'
 URL_ROOT = ''
+
+code_evaluators = {"python": "python_code_evaluator.PythonCodeEvaluator",
```
Comment (prabhuramachandran) [inline testapp/exam/settings.py]: Shouldn't this be absolute imports and not relative imports?

File: testapp/test_bash_evaluation.py [hunk 32]
```diff
@@ -0,0 +1,42 @@
+
```
Comment (prabhuramachandran) [inline testapp/test_bash_evaluation.py]: Move all the tests into a separate "tests" directory in the appropriate place with an `__init__.py` also in the directory.

Comment (ankitjavalkar) [inline testapp/test_bash_evaluation.py]: Done

File: testapp/exam/settings.py (hunk 31 above)
Comment (ankitjavalkar) [inline testapp/exam/settings.py]: I tried using absolute paths - But since these evaluator paths will be imported within `code_server.py` and eventually `language_registry.py` which are located within `exam` directory, hence such absolute paths don't work.

File: testapp/exam/python_code_evaluator.py (hunk 30 above)
Comment (ankitjavalkar) [inline testapp/exam/python_code_evaluator.py]: Done

File: testapp/exam/models.py (hunk 29 above)
Comment (ankitjavalkar) [inline testapp/exam/models.py]: Done

File: testapp/exam/java_code_evaluator.py (hunk 28 above)
Comment (ankitjavalkar) [inline testapp/exam/java_code_evaluator.py]: Done

File: testapp/exam/cpp_code_evaluator.py (hunk 23 above)
Comment (ankitjavalkar) [inline testapp/exam/cpp_code_evaluator.py]: Done

File: testapp/exam/code_evaluator.py (hunk 19 above)
Comment (ankitjavalkar) [inline testapp/exam/code_evaluator.py]: Done

File: testapp/exam/code_evaluator.py (hunk 20 above)
Comment (ankitjavalkar) [inline testapp/exam/code_evaluator.py]: Done

File: testapp/exam/code_evaluator.py (hunk 21 above)
Comment (ankitjavalkar) [inline testapp/exam/code_evaluator.py]: Done

File: testapp/exam/code_server.py (hunk 22 above)
Comment (ankitjavalkar) [inline testapp/exam/code_server.py]: Done

**Comment by prabhuramachandran** [comment]
Thanks for all the changes, looks good now, merging.  The test suite needs to run regularly, I suggest setting up travis-ci for this.
